### PR TITLE
fix(`forge`): fix cache search for verification

### LINKED ISSUE #7042: Duplicate contract names error for contracts with different file names
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (2cf84d9 2024-02-07T00:16:06.769732275Z)

### What command(s) is the bug in?

forge create

### Operating System

Linux

### Describe the bug

While running forge create I receive an error saying I have duplicate contract names. While that is true I have them in two separate, distinctly named files - `ModifiedTokenDistro.sol` and `TokenDistro.sol`.
image

When I run this command: 
```
forge create --private-key $PRIVATE_KEY --etherscan-api-key $GNOSISSCAN_API_KEY --chain-id 100 --verify --rpc-url $GNOSIS_RPC_URL --verifier-url https://api.gnosisscan.io/api contracts/ModifiedTokenDistro.sol:TokenDistro

```
I get this error: 
```
Error: 
If cache is disabled, compiler version must be either provided with `--compiler-version` option or set in foundry.toml

Context:
- contract with duplicate name `TokenDistro`. please pass the path instead
```

image

However I did specify the relative path to the contract file and the contract name - so why does it think there is a duplicate contract if I did explicitly specify the file path? 

here is a public URL to the contract in question
https://github.com/Giveth/giveconomy-foundry/blob/modify-allocation-functions/contracts/ModifiedTokenDistro.sol

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Closes #7042

## Solution

Current `EtherscanVerificationProvider::cache_entry` always looks for contract in cache by name, even if path present, this PR fixes it and uses path where possible.

### CODE REVIEW

File: crates/forge/bin/cmd/verify/etherscan/mod.rs [hunk 1]
```diff
@@ -212,15 +214,27 @@ impl EtherscanVerificationProvider {
...
+                (
+                    path.clone(),
+                    cache
+                        .entry(&path)
+                        .ok_or_eyre(format!("Cache entry not found for {}", path.display()))?
+                        .to_owned(),
+                )
+            }
+            None => get_cached_entry_by_name(&cache, &contract.name)?,
+        };
```
Comment (mattsse) [inline crates/forge/bin/cmd/verify/etherscan/mod.rs:236]: style: this should be `if let some else`

File: crates/forge/bin/cmd/verify/etherscan/mod.rs [hunk 2]
```diff
@@ -350,7 +364,7 @@ impl EtherscanVerificationProvider {
         let path = match args.contract.path.as_ref() {
             Some(path) => project.root().join(path),
             None => {
```
Comment (mattsse) [inline crates/forge/bin/cmd/verify/etherscan/mod.rs:366]: while we're at it, let's also do if let some else for this

File: crates/forge/bin/cmd/verify/etherscan/mod.rs [hunk 3]
```diff
@@ -212,15 +214,27 @@ impl EtherscanVerificationProvider {
...
-        let (path, entry) = get_cached_entry_by_name(&cache, contract_name)?;
-        let contract: CompactContract = cache.read_artifact(path.clone(), contract_name)?;
+        let (path, entry) = match contract.path.as_ref() {
+            Some(path) => {
+                let path = project.root().join(path);
+                (
+                    path.clone(),
+                    cache
+                        .entry(&path)
+                        .ok_or_eyre(format!("Cache entry not found for {}", path.display()))?
```
Comment (mattsse) [inline crates/forge/bin/cmd/verify/etherscan/mod.rs:231]: TIL ok_or_eyre

**Review by mattsse — APPROVED** [review]
nice!

pedantic style nit

File: crates/forge/bin/cmd/verify/etherscan/mod.rs [hunk 4]
```diff
@@ -212,15 +214,26 @@ impl EtherscanVerificationProvider {
...
         let cache = project.read_cache_file()?;
-        let (path, entry) = get_cached_entry_by_name(&cache, contract_name)?;
-        let contract: CompactContract = cache.read_artifact(path.clone(), contract_name)?;
+        let (path, entry) = if let Some(path) = contract.path.as_ref() {
+            let path = project.root().join(path);
+            (
+                path.clone(),
+                cache
+                    .entry(&path)
+                    .ok_or_eyre(format!("Cache entry not found for {}", path.display()))?
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/verify/etherscan/mod.rs:230]:
```suggestion
                    .ok_or_else(|| eyre::eyre!("Cache entry not found for {}", path.display()))?
```

**Review by DaniPopes — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

**Pull request merged**